src/bluetooth_linux.c

```diff
@@ -0,0 +1,673 @@
+				return 0;
+
+			if (value_len > dev->controlpoint_size)
+				value_len = dev->controlpoint_size;
+
+			memcpy(dev->reply_buf, value, value_len);
+			dev->reply_len = value_len;
+
+			return 1;
+		} else {
+			sd_bus_message_skip(m, "v");
+		}
+
+		sd_bus_message_exit_container(m);
+	}
+	return 0;
+}
+
+static void iterate_over_all_objs(sd_bus_message *reply,
+				  void (*new_dbus_interface)(void *,
+				  const char *,
+				  sd_bus_message *), void *data)
+{
+	if (sd_bus_message_enter_container(reply, SD_BUS_TYPE_ARRAY, "{oa{sa{sv}}}") <= 0)
+		return;
+
+	while (0 < sd_bus_message_enter_container(reply, SD_BUS_TYPE_DICT_ENTRY, "oa{sa{sv}}")) {
+		const char *ifacepath = NULL;
+		if (sd_bus_message_read_basic(reply, 'o', &ifacepath) <= 0)
+			return;
+
+		if (sd_bus_message_enter_container(reply, SD_BUS_TYPE_ARRAY, "{sa{sv}}") < 0)
+			return;
+		while (0 < sd_bus_message_enter_container(reply, SD_BUS_TYPE_DICT_ENTRY, "sa{sv}")) {
+			new_dbus_interface(data, ifacepath, reply);
+			sd_bus_message_exit_container(reply);
+		}
+		sd_bus_message_exit_container(reply);
+		sd_bus_message_exit_container(reply);
+	}
+}
+
+void *
+fido_bluetooth_open(const char *path)
+{
+	struct ble *newdev;
+	sd_bus_message *reply = NULL;
+	int ret;
+	if (!fido_is_bluetooth(path))
+		return NULL;
+
+	path += strlen(FIDO_BLUETOOTH_PREFIX);
+
+	newdev = calloc(1, sizeof(*newdev));
+	if (!newdev)
+		return NULL;
+
+	newdev->paths.dev = strdup(path);
+	if (!newdev->paths.dev)
+		goto out;
+
+	if (0 > sd_bus_default_system(&newdev->bus))
+		goto out;
+
+	if (0 > sd_bus_call_method(newdev->bus, "org.bluez",
+		path, "org.freedesktop.DBus.Properties", "GetAll", NULL, &reply,
+		"s", DBUS_DEV_IFACE))
+		goto out;
+
+	if (!ble_fido_is_useable_device(DBUS_DEV_IFACE, reply, false, NULL))
+		goto out;
+
+	sd_bus_message_unref(reply);
+	reply = NULL;
+	ret = sd_bus_call_method(newdev->bus, "org.bluez", "/", "org.freedesktop.DBus.ObjectManager", "GetManagedObjects", NULL, &reply, "");
+	if (ret <= 0)
+		goto out;
+
+	sd_bus_message_rewind(reply, 1);
+	iterate_over_all_objs(reply, collect_device_chars, newdev);
+
+	sd_bus_message_unref(reply);
+	reply = NULL;
+
+	if (newdev->paths.status &&
+	    newdev->paths.control_point &&
+	    newdev->paths.control_point_length &&
+	    newdev->paths.service_revision) {
+		uint8_t cp_len[2];
+		uint8_t revision;
+		if (0 > sd_bus_call_method(newdev->bus, "org.bluez", newdev->paths.control_point_length,
+					DBUS_CHAR_IFACE, "ReadValue", NULL, &reply, "a{sv}", 0))
+			goto out;
+
+		if (0 > sd_bus_message_read(reply, "ay", 2, cp_len, cp_len + 1))
+			goto out;
+
+		sd_bus_message_unref(reply);
+		reply = NULL;
+
+		if (0 > sd_bus_call_method(newdev->bus, "org.bluez", newdev->paths.service_revision,
+					DBUS_CHAR_IFACE, "ReadValue", NULL, &reply, "a{sv}", 0))
+			goto out;
+		if (0 > sd_bus_message_read(reply, "ay", 1, &revision))
+			goto out;
+
+		/* for simplicity, we allow now only FIDO2 */
+		if (!(revision & 0x20))
```

Comment:
The specification says

>  The client SHALL write a value to this characteristic with exactly 1 bit set before sending any FIDO commands unless u2fServiceRevision is present and U2F 1.0 compatibility is desired

which I believe we are missing?

Reply:
yes, it is missing